stratum/hal/lib/tdi/tdi_soft_error.h

```diff
@@ -0,0 +1,56 @@
+// Copyright 2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef STRATUM_HAL_LIB_TDI_TDI_SOFT_ERROR_H_
+#define STRATUM_HAL_LIB_TDI_TDI_SOFT_ERROR_H_
+
+#include "stratum/glue/status/status.h"
+#include "stratum/hal/lib/tdi/tdi_status.h"
+#include "stratum/public/lib/error.h"
+
+namespace stratum {
+namespace hal {
+namespace tdi {
+
+// Checks whether a TDI status code is a soft error.
+static inline bool IsSoftTdiError(tdi_status_t err) {
+  return (err == TDI_ALREADY_EXISTS) || (err == TDI_OBJECT_NOT_FOUND) ||
```

Comment:
What is the definition of a soft error in this implementation? Why is "object/table not found" considered a soft error?

Reply:
1. When I reviewed the updated HSD-ES ticket, I discovered that it contained log messages like this:

   ```text
   E20231201 00:28:53.970028 5140 tdi_table_manager.cc:624] Return Error: tdi_sde_interface_->GetTableEntry(device_, session, table_id, table_key.get(), table_data.get()) failed with StratumErrorSpace::ERR_ENTRY_NOT_FOUND: 'table->entryGet(*real_session->tdi_session_, *dev_tgt, flags, *real_table_key->table_key_, real_table_data->table_data_.get())' failed with error message: Object not found. 
   E20231201 00:28:53.970144 5140 es2k_node.cc:278] StratumErrorSpace::ERR_AT_LEAST_ONE_OPER_FAILED: One or more read operations failed.
   E20231201 00:28:53.970456 5140 p4_service.cc:353] Failed to read forwarding entries from node 1: One or more read operations failed.
   ```
2. The commentary showed that they were also bothered about these issues (i.e., my previous commit had addressed only half their problem).
3. ALREADY_EXISTS and NOT_FOUND are two sides of the same coin. The conditions are easily recoverable, which is what makes them "soft" errors.
4. The customer is complaining about the fact that the intermediate code is logging, not just ALREADY_EXISTING, but ALREADY_EXISTING and NOT_FOUND. They are therefore complaining about our logging soft errors.
5. When TDI status codes are converted to Stratum status codes, TDI_TABLE_NOT_FOUND and TDI_OBJECT_NOT_FOUND are both mapped to ERR_ENTRY_NOT_FOUND. I therefore treat them as identical.
6. The idea is to reduce the amount of special-purpose code by treating all errors with this classification identically.

Reply:
Point taken. I've added some additional documentation to the header file.

Reply:
More to the point, there is code (in ovsp4rt, and perhaps elsewhere) that checks for the existence of a table entry before it proceeds. TDI does not provide an EntryExists() method, so the code does a read and checks the status code.

And think about it. If "entry already exists" is a soft error, why wouldn't "entry does not exist" be a soft error? They're complementary.

Reply:
The chief effect of this change is to go from saying that a _specific_ status code (ALREADY_EXISTS) deserves special handling to saying that there's a _class_ of status codes that deserve special handling, and ALREADY_EXISTS and NOT_FOUND are both members.

The other effect is to call these _soft errors_ instead of appending "(may not be an error)" to the message text. I suppose I could have changed the text to "(note that the application may not consider this to be an error)", but calling it a "soft error" is more elegant.